lib/gauge_plaq.cu

```diff
@@ -1,52 +1,34 @@
-#include <tune_quda.h>
 #include <gauge_field.h>
-#include <jitify_helper.cuh>
-#include <kernels/gauge_plaq.cuh>
 #include <instantiate.h>
+#include <tunable_reduction.h>
+#include <kernels/gauge_plaq.cuh>
 
 namespace quda {
 
   template<typename Float, int nColor, QudaReconstructType recon>
-  class GaugePlaq : TunableLocalParityReduction {
+  class GaugePlaq : public TunableReduction2D<> {
     const GaugeField &u;
     double2 &plq;
 
   public:
     GaugePlaq(const GaugeField &u, double2 &plq) :
+      TunableReduction2D(u),
       u(u),
       plq(plq)
     {
-#ifdef JITIFY
-      create_jitify_program("kernels/gauge_plaq.cuh");
-#endif
-      strcpy(aux, compile_type_str(u));
-      apply(0);
+      apply(device::get_default_stream());
     }
 
-    void apply(const qudaStream_t &stream){
-      if (u.Location() == QUDA_CUDA_FIELD_LOCATION) {
-	TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
-        GaugePlaqArg<Float, nColor, recon> arg(u);
-#ifdef JITIFY
-        using namespace jitify::reflection;
-        jitify_error = program->kernel("quda::computePlaq")
-          .instantiate((int)tp.block.x,type_of(arg))
-          .configure(tp.grid,tp.block,tp.shared_bytes,stream).launch(arg);
-        arg.launch_error = jitify_error == CUDA_SUCCESS ? QUDA_SUCCESS : QUDA_ERROR;
-#else
-	LAUNCH_KERNEL_LOCAL_PARITY(computePlaq, (*this), tp, stream, arg, decltype(arg));
-#endif
-        arg.complete(plq);
-        if (!activeTuning()) {
-          comm_allreduce_array((double*)&plq, 2);
-          for (int i = 0; i < 2; i++) ((double*)&plq)[i] /= 9.*2*arg.threads*comm_size();
-        }
-      } else {
-	errorQuda("CPU not supported yet\n");
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      GaugePlaqArg<Float, nColor, recon> arg(u);
+      launch<Plaquette>(plq, tp, stream, arg);
+      if (!activeTuning()) {
+        for (int i = 0; i < 2; i++) ((double*)&plq)[i] /= 9.*2*arg.threads.x*comm_size();
```

Comment:
C-style cast

Reply:
Best thing to do here is just use a `std::vector` instead of the `double2`.  Also, there's no need for the 
```c++
if (!activeTuning())
``` 

guard here.  Fixed in 6b0981b45

Reply:
`std::array<double,2>` ?

Reply:
@weinbe2 the `launch` interface can accept either an array pointer or a `std::vector`.  Hence why in my revision I've switched to using `std::vector` to remove the ugly cast.

Reply:
Are you receptive it letting it also accept an `std::array<T,N>`? I think it'd be a good practice to get into using `std::array` in cases where we know the size is fixed.

Reply:
It can be a post-GK addition if needed.  No need to add this now and I think a more pressing need at the moment I think would more removal of passing pointers for arrays in host code and replacing those with `std::vector`.